source/common/common/utility.h

```diff
@@ -808,6 +964,60 @@ class InlineString : public InlineStorage {
   char data_[];
 };
 
+class TinyString {
+public:
+  TinyString() = default;
+  TinyString(absl::string_view str) { populateFromStringView(str); }
+  TinyString(TinyString&& src) : data_(std::move(src.data_)) { src.data_.reset(); }
+  TinyString(const TinyString& src) {
+    if (src.hasValue()) {
+      populateFromTinyString(src);
+    }
+  }
+
+  TinyString& operator=(const TinyString& src) {
+    if (&src != this) {
+      if (src.hasValue()) {
+        populateFromTinyString(src);
+      } else {
+        data_.reset();
+      }
+    }
+    return *this;
+  }
+
+  void operator=(TinyString&& src) {
+    if (&src != this) {
+      data_ = std::move(src.data_);
+    }
+  }
+
+  /**
+   * Caller must ensure hasValue() is true.
+   *
+   * @return a std::string copy of the TinyString
+   */
+  std::string toString() const { return std::string(toStringView()); }
```

Comment:
Is this method used (or expected to be used) outside of new tests in this PR? It hides the cost of allocation and copy. Maybe a better approach is not to provide it at all, and make the called use the `toStringView` to construct std::string if they really need it, making the cost of doing so more explicit.

Reply:
that's OK with me also, but to me, it seems obvious that a function with signature `std::string toString() const` must do a copy.